docs/proposals/20201020-management-cluster-operator.md

```diff
@@ -0,0 +1,844 @@
+  new provider is created, the operator will determine the version to use
+  applying the same rules implemented in clusterctl (latest).
+- The content of `ProviderSpec.SecretName` will be treated as immutable. If
+  changes need to be made, a new object can be created and the name should be
+  updated.
+- The content of `ProviderSpec.SecretName` should be in the form of key:value.
+- The `ProviderSpec.SecretName` specified must be in the same namespace as the
+  provider.
+- When a provider will be paused the number of replicas will be scaled to 0;
+  the operator adds a new
+  `management.cluster.x-k8s.io/original-controller-replicas` annotations to
+  store the original replica count.
+
+**Validation rules for ProviderSpec.FetchConfiguration**
+- If the FetchConfiguration is null, then the operator will try to apply the
+  embedded fetch configuration for the given kind and `ObjectMeta.Name`. For
+  example, the infrastructure name `aws` will fetch artifacts from
+  https://github.com/kubernetes-sigs/cluster-api-provider-aws/releases.
+- If FetchConfiguration is not nil, exactly one of `URL` or `Selector` must be
+  specified.
+- `FetchConfiguration.Selector` is used to fetching provider’s components and
+  metadata from ConfigMaps stored inside the cluster. Each ConfigMap is
+  expected to contain components and metadata for a specific version only. So
+  if multiple versions of the providers need to be specified, they can be
+  added as separate ConfigMaps and labeled with the same selector. This
+  emulates the “local” provider repositories but now it’s within the
+  management cluster.
+
+**Validation Rules for ProviderSpec.ManagerSpec**
+- The ControllerManagerConfigurationSpec is a type from
+  `controller-runtime/pkg/config` and is an embedded into the `ManagerSpec`.
+  This type will expose LeaderElection, SyncPeriod, Webhook, Health and
+  Metrics configurations.
+- If `ManagerSpec.Debug` is set to true, the operator will not allow changes
+  to other properties since it is in Debug mode.
+- If you need to set specific concurrency values for each reconcile loop (e.g.
+  `awscluster-concurrency`), you can leave
+  `ManagerSpec.MaxConcurrentReconciles` nil and use `Container.Args`.
+- If `ManagerSpec.MaxConcurrentReconciles` is set and a specific concurrency
+  flag such as `awscluster-concurrency` is set on the `Container.Args`, then
+  the more specific concurrency flag will have higher precedence.
+
+
+**Validation Rules for ContainerSpec**
+- The `ContainerSpec.Args` will ignore the key `namespace` since the operator
+  enforces a deployment model where all the providers should be configured to
+  watch all the namespaces.
+- Explicit controller manager properties defined in the `Provider.ManagerSpec`
+  will have higher precedence than those defined in `ContainerSpec.Args`. That
+  is, if `ManagerSpec.SyncPeriod`  is defined it will be used instead of the
+  container arg `sync-period`.
```

Comment:
That is one option but it feels restrictive because now the user will need to look up the API and see what they can define in the ManagerSpec vs what they can define in the ContainerSpec.Args.
I thought that it would be better to just have this rule so that we can just do a union of all the flags that need to be set on the controller manager and use the one defined in ManagerSpec if there are duplicates.

Reply:
I have no strong opinions, but I personally prefer to get an explicit validation error message when creating the object instead of trying to find out ex-post what is the merge rule